agent/csi/volumes.go

```diff
@@ -0,0 +1,259 @@
+		pluginMap: make(map[string]*NodePlugin),
+	}
+}
+
+// Get returns a volume published path for the provided volume ID.  If the volume doesn't exist, returns empty string.
+func (r *volumes) Get(volumeID string) (string, error) {
+	r.mapMu.RLock()
+	defer r.mapMu.RUnlock()
+	if plugin, ok := r.pluginMap[volumeID]; ok {
+		return plugin.GetPublisedPath(volumeID)
+	}
+	return "", nil
+}
+
+// Add adds one or more volumes to the volume map.
+func (r *volumes) Add(volumes ...api.VolumeAssignment) {
+	r.mu.Lock()
+	var volumeObjects []*api.VolumeAssignment
+	defer r.mu.Unlock()
+	ctx := context.Background()
+	for _, volume := range volumes {
+		v := volume.Copy()
+		log.G(ctx).WithField("method", "(*volumes).Add").Debugf("Add Volume:%v", volume.VolumeID)
+
+		r.m[volume.VolumeID] = v
+		driverName := v.Driver.Name
+		r.mapMu.Lock()
+		if _, ok := r.pluginMap[driverName]; !ok {
+			r.pluginMap[driverName] = NewNodePlugin(driverName)
+		}
+		r.mapMu.Unlock()
+		volumeObjects = append(volumeObjects, v)
+	}
+	go r.iterateVolumes(volumeObjects, true)
+}
+
+func (r *volumes) addVolumes(volumeObjects []*api.VolumeAssignment) {
+	ctx := context.Background()
+	for _, v := range volumeObjects {
+		go r.tryAddVolume(ctx, v)
+	}
+}
+
+func (r *volumes) iterateVolumes(volumeObjects []*api.VolumeAssignment, isAdd bool) {
+	ctx := context.Background()
+	for _, v := range volumeObjects {
+		if isAdd {
+			go r.tryAddVolume(ctx, v)
+		} else {
+			go r.tryRemoveVolume(ctx, v)
+		}
+	}
+}
+
+func (r *volumes) tryAddVolume(ctx context.Context, assignment *api.VolumeAssignment) {
+	r.tryMu.Lock()
+	defer r.tryMu.Unlock()
+
+	r.mapMu.RLock()
+	plugin, ok := r.pluginMap[assignment.VolumeID]
+	r.mapMu.RUnlock()
+	if !ok {
+		log.G(ctx).Debugf("plugin not found for VolumeID:%v", assignment.VolumeID)
+		return
+	}
+	if err := plugin.NodeStageVolume(ctx, assignment); err != nil {
+		waitFor := initialBackoff
+	retryStage:
+		for i := 0; i < maxRetries; i++ {
+			select {
+			case <-ctx.Done():
+				// selecting on ctx.Done() allows us to bail out of retrying early
+				return
+			case <-time.After(waitFor):
+				// time.After is better than using time.Sleep, because it blocks
+				// on a channel read, rather than suspending the whole
+				// goroutine. That lets us do the above check on ctx.Done().
+				//
+				// time.After is convenient, but it has a key problem: the timer
+				// is not garbage collected until the channel fires. this
+				// shouldn't be a problem, unless the context is canceled, there
+				// is a very long timer, and there are a lot of other goroutines
+				// in the same situation.
+				if err := plugin.NodeStageVolume(ctx, assignment); err == nil {
+					break retryStage
+				}
+			}
+			// if the exponential factor is 2, you can avoid using floats by
+			// doing bit shifts. each shift left increases the number by a power
+			// of 2. we can do this because Duration is ultimately int64.
+			waitFor = waitFor << 1
```

Comment:
100ms, when bit shifted 25 times, is a really, really big number. Like, 38 days big. You should cap the maximum time between retries to something like an hour.